File: python/clustering.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score, normalized_mutual_info_score, adjusted_rand_score
from sklearn.decomposition import PCA
from scipy.cluster.hierarchy import dendrogram, linkage
import warnings
from matplotlib.backends.backend_pdf import PdfPages

# Configuration
warnings.filterwarnings('ignore')
plt.rcParams['figure.figsize'] = (10, 6)
sns.set_style('whitegrid')

# Ensure visuals directory exists inside python folder
VISUALS_DIR = os.path.join(os.path.dirname(__file__), 'visuals')
if not os.path.exists(VISUALS_DIR):
    os.makedirs(VISUALS_DIR)

def save_plot(filename):
    """Saves the current plot to the visuals directory and closes it."""
    path = os.path.join(VISUALS_DIR, filename)
    plt.savefig(path)
    plt.close()

def fix_invalid_date(date_str):
    """Tries to fix invalid dates by decreasing the day by 1."""
    try:
        return pd.to_datetime(date_str)
    except:
        try:
            date_str = str(date_str)
            if '/' in date_str:
                parts = date_str.split('/')
                month, day, year = int(parts[0]), int(parts[1]), int(parts[2])
            elif '-' in date_str:
                parts = date_str.split('-')
                year, month, day = int(parts[0]), int(parts[1]), int(parts[2])
            else:
                return pd.NaT
            
            while day > 0:
                try:
                    return pd.Timestamp(year=year, month=month, day=day)
                except:
                    day -= 1
        except:
            pass
        return pd.NaT

def main():
    print("Loading dataset...")
    try:
        # Load dataset relative to the script location
        script_dir = os.path.dirname(os.path.abspath(__file__))
        file_path = os.path.join(script_dir, '../project cluster.xlsx')
        df = pd.read_excel(file_path)
    except FileNotFoundError:
        print(f"Error: '{file_path}' not found. Please check the file path.")
        return
    except Exception as e:
        print(f"Error loading file: {e}")
        return
    
    
    # --- Preprocessing ---
    print("Preprocessing data...")
    df_processed = df.copy()
    
    # Feature Engineering
    df_processed['Total_Guests'] = df_processed['number.of.adults'] + df_processed['number.of.children']
    df_processed['Is_Family'] = (df_processed['number.of.children'] > 0).astype(int)
    df_processed['Total_Nights'] = df_processed['number.of.weekend.nights'] + df_processed['number.of.week.nights']
    df_processed['Cancellation_Ratio'] = (df_processed['P.C'] + 1) / (df_processed['P.not.C'] + df_processed['P.C'] + 2)
    df_processed['Price_per_Person'] = df_processed['average.price'] / df_processed['Total_Guests']
    df_processed['Price_per_Person'].replace([np.inf, -np.inf], np.nan, inplace=True)
    
    # Date fixing - R script drops it, so we drop it too. 
    # (R: df <- df %>% select(-date.of.reservation))
    if 'date.of.reservation' in df_processed.columns:
        df_processed.drop(columns=['date.of.reservation'], inplace=True)
    
    # Value capping/binning - Matching R script logic
    df_processed['number.of.children'] = df_processed['number.of.children'].apply(lambda x: 0 if x == 0 else 1)
    df_processed['number.of.weekend.nights'] = df_processed['number.of.weekend.nights'].apply(lambda x: min(x, 3))
    df_processed['number.of.week.nights'] = df_processed['number.of.week.nights'].apply(lambda x: min(x, 6))
    df_processed['P.C'] = df_processed['P.C'].apply(lambda x: 0 if x == 0 else 1)
    df_processed['P.not.C'] = df_processed['P.not.C'].apply(lambda x: 0 if x == 0 else 1)
    df_processed['Total_Guests'] = df_processed['Total_Guests'].apply(lambda x: 1 if x == 0 else min(x, 3))
    df_processed['Total_Nights'] = df_processed['Total_Nights'].round().astype(int).apply(lambda x: min(x, 8))
    
    # Save booking status and drop unused columns
    booking_status = df_processed['booking.status'].copy()
    df_processed.drop(columns=['Booking_ID', 'booking.status'], inplace=True)
    
    # One-hot encoding
    categorical_features = ['market.segment.type']
    # drop_first=True will drop the first alphabetical category. 
    # If 'Aviation' is present and first, it will be dropped, matching R's logic.
    df_encoded = pd.get_dummies(df_processed, columns=categorical_features, drop_first=True)
    
    # Drop meal, room type, week related columns
    cols_to_drop = [col for col in df_encoded.columns if col.startswith(('type.of.meal', 'room.type'))] 
    df_encoded.drop(columns=cols_to_drop, inplace=True)
    
    # Fill NaNs
    if df_encoded.isnull().sum().sum() > 0:
        df_encoded = df_encoded.fillna(df_encoded.mean())
        
    # Scaling (0 to 1) - Changed from (-1, 1) to match R
    scaler = MinMaxScaler(feature_range=(0, 1))
    df_scaled_array = scaler.fit_transform(df_encoded)
    df_scaled = pd.DataFrame(df_scaled_array, columns=df_encoded.columns)
    
    print("Data processed and scaled.")
    
    # --- Plotting Distributions ---
    print("Generating distribution plots...")
    numeric_cols = [
        'Price_per_Person', 'Cancellation_Ratio', 'Total_Nights', 'Is_Family', 
        'Total_Guests', 'special.requests', 'average.price', 'P.not.C', 'P.C', 
        'repeated', 'lead.time', 'car.parking.space', 'number.of.week.nights', 
        'number.of.weekend.nights', 'number.of.children', 'number.of.adults'
    ]
    
    for col in numeric_cols:
        if col in df_scaled.columns:
            plt.figure(figsize=(10, 6))
            plt.hist(df_scaled[col], bins=30, edgecolor='black', alpha=0.7)
            plt.title(f'Distribution (Scaled): {col}')
            plt.xlabel('Scaled Value [0, 1]')
            plt.ylabel('Frequency')
            plt.grid(axis='y', alpha=0.3)
            plt.tight_layout()
            save_plot(f'dist_{col}.png')

    # --- Clustering ---
    print("Running clustering algorithms...")
    X = df_scaled.values
    k_range = range(2, 11)
    
    kmeans_silhouette = []
    hierarchical_silhouette = []
    nmi_scores = []
    ari_scores = []
    
    kmeans_labels_dict = {}
    hierarchical_labels_dict = {}
    kmeans_inertia = []

    # 1. K-Means Loop
    print("\nΕκτέλεση K-Means για k = 2 έως 10...")
    for k in k_range:
        kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
        k_labels = kmeans.fit_predict(X)
        kmeans_labels_dict[k] = k_labels
        kmeans_inertia.append(kmeans.inertia_)
        s_score = silhouette_score(X, k_labels)
        kmeans_silhouette.append(s_score)
        print(f"K-Means με k={k}... ✓ (Inertia: {kmeans.inertia_:.2f}, Silhouette: {s_score:.4f})")
    print("K-Means ολοκληρώθηκε!")

    # 2. Hierarchical Loop
    print("\nΕκτέλεση Hierarchical Clustering για k = 2 έως 10...")
    
    # Dendrogram (using a sample for visibility)
    print("Generating Dendrogram...")
    plt.figure(figsize=(12, 8))
    # Use a sample for dendrogram to avoid overcrowding
    sample_idx = np.random.choice(X.shape[0], min(5000, X.shape[0]), replace=False)
    X_sample = X[sample_idx]
    Z = linkage(X_sample, method='ward')
    dendrogram(Z, truncate_mode='lastp', p=50, leaf_rotation=90., leaf_font_size=10., show_contracted=True)
    plt.title('Hierarchical Clustering Dendrogram (Ward Linkage, Sample=5000)')
    plt.xlabel('Cluster Size')
    plt.ylabel('Distance')
    plt.tight_layout()
    save_plot('dendrogram.png')
    print("Dendrogram saved.")

    for k in k_range:
        hierarchical = AgglomerativeClustering(n_clusters=k, linkage='ward')
        h_labels = hierarchical.fit_predict(X)
        hierarchical_labels_dict[k] = h_labels
        s_score = silhouette_score(X, h_labels)
        hierarchical_silhouette.append(s_score)
        print(f"Hierarchical με k={k}... ✓ (Silhouette: {s_score:.4f})")
    print("Hierarchical Clustering ολοκληρώθηκε!")

    # 3. Comparison Loop
    print("\nΥπολογισμός NMI και ARI μεταξύ K-Means και Hierarchical...")
    for k in k_range:
        k_labels = kmeans_labels_dict[k]
        h_labels = hierarchical_labels_dict[k]
        nmi = normalized_mutual_info_score(k_labels, h_labels)
        ari = adjusted_rand_score(k_labels, h_labels)
        nmi_scores.append(nmi)
        ari_scores.append(ari)
        print(f"k={k}: NMI={nmi:.4f}, ARI={ari:.4f}")
    print("Υπολογισμοί ολοκληρώθηκαν!")
        
    # Evaluation Plot
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))
    k_values = list(k_range)
    
    axes[0, 0].plot(k_values, kmeans_inertia, 'o-')
    axes[0, 0].set_title('K-Means Inertia')
    
    axes[0, 1].plot(k_values, kmeans_silhouette, 'o-', label='K-Means')
    axes[0, 1].plot(k_values, hierarchical_silhouette, 's-', label='Hierarchical')
    axes[0, 1].legend()
    axes[0, 1].set_title('Silhouette Score')
    
    axes[1, 0].plot(k_values, nmi_scores, 'o-')
    axes[1, 0].set_title('NMI Score')
    
    axes[1, 1].plot(k_values, ari_scores, 'o-')
    axes[1, 1].set_title('ARI Score')
    
    plt.tight_layout()
    save_plot('clustering_evaluation.png')
    
    # --- Feature Elimination (k=4) ---
    print("Running feature elimination analysis (k=4)...")
    k = 4
    baseline_silhouette = silhouette_score(X, kmeans_labels_dict[4])
    
    results = []
    for i, feature in enumerate(df_scaled.columns):
        X_reduced = np.delete(X, i, axis=1)
        kmeans_reduced = KMeans(n_clusters=k, random_state=42, n_init=10)
        labels_reduced = kmeans_reduced.fit_predict(X_reduced)
        score = silhouette_score(X_reduced, labels_reduced)
        results.append({'feature': feature, 'improvement': score - baseline_silhouette})
    
    sorted_results = sorted(results, key=lambda x: x['improvement'], reverse=True)
    print("Top 5 features to remove (improve silhouette):")
    for res in sorted_results[:5]:
        print(f"  {res['feature']}: +{res['improvement']:.4f}")

    # --- Final Clustering (k=4) ---
    print("Finalizing clusters (k=4)...")
    k_optimal = 4
    kmeans_optimal = KMeans(n_clusters=k_optimal, random_state=42, n_init=10)
    cluster_labels = kmeans_optimal.fit_predict(X)
    
    # PCA Visualization
    print("Generating PCA plot...")
    pca = PCA(n_components=2)
    X_pca = pca.fit_transform(X)
    
    plt.figure(figsize=(10, 8))
    scatter = plt.scatter(X_pca[:, 0], X_pca[:, 1], c=cluster_labels, cmap='viridis', alpha=0.6, s=10)
    plt.title(f'K-Means Clustering (k={k_optimal}) - PCA Projection')
    plt.xlabel(f'Principal Component 1 ({pca.explained_variance_ratio_[0]:.2%} variance)')
    plt.ylabel(f'Principal Component 2 ({pca.explained_variance_ratio_[1]:.2%} variance)')
    plt.colorbar(scatter, label='Cluster')
    plt.grid(True, alpha=0.3)
    save_plot('pca_clusters.png')
    print("PCA plot saved.")
    
    df_with_clusters = df_encoded.copy()
    df_with_clusters['Cluster'] = cluster_labels
    df_with_clusters['booking.status'] = booking_status.values
    
    # Cluster Analysis Plots
    variables_to_analyze = [
        'number.of.adults', 'number.of.children', 'car.parking.space', 
        'lead.time', 'repeated', 'P.C', 'P.not.C', 'average.price', 
        'special.requests', 'Total_Guests', 'Is_Family', 'Total_Nights', 
        'Cancellation_Ratio', 'Price_per_Person', 'number.of.week.nights', 
        'number.of.weekend.nights'
    ]
    market_segment_cols = [c for c in df_with_clusters.columns if 'market.segment.type' in c]
    
    cluster_means = df_with_clusters.groupby('Cluster')[variables_to_analyze + market_segment_cols].mean()
    
    # Bar plots for variables
    for var in variables_to_analyze + market_segment_cols:
        plt.figure(figsize=(10, 6))
        plt.bar(cluster_means.index, cluster_means[var], color=['steelblue', 'coral', 'lightgreen', 'gold'])
        plt.title(f'Cluster Comparison: {var}')
        save_plot(f'cluster_comp_{var}.png')
        
    # Summary Table
    cluster_summary = cluster_means.copy()
    cluster_summary['Count'] = df_with_clusters.groupby('Cluster').size()
    cols_order = ['Count'] + variables_to_analyze + market_segment_cols
    cluster_summary = cluster_summary[cols_order]
    
    print("\nCluster Summary (Means):")
    print(cluster_summary)

    
    # Heatmap
    plt.figure(figsize=(16, 6))
    scaler_heatmap = MinMaxScaler()
    cluster_summary_norm = pd.DataFrame(
        scaler_heatmap.fit_transform(cluster_summary),
        columns=cluster_summary.columns,
        index=cluster_summary.index
    )
    sns.heatmap(cluster_summary_norm.T, annot=cluster_summary_norm.T.round(2), cmap='RdYlGn')
    plt.title('Cluster Means Heatmap (Normalized)')
    plt.tight_layout()
    save_plot('cluster_heatmap.png')
    
    # Cancellation Rate Plot
    cancellation_rates = df_with_clusters.groupby('Cluster')['booking.status'].apply(
        lambda x: (x == 'Canceled').mean() * 100
    )
    
    plt.figure(figsize=(10, 6))
    plt.bar(cancellation_rates.index, cancellation_rates, color=['steelblue', 'coral', 'lightgreen', 'gold'])
    plt.title('Cancellation Rate by Cluster')
    plt.ylabel('Cancellation Rate (%)')
    save_plot('cancellation_rate.png')
    
    print("\nCancellation Rates:")
    print(cancellation_rates)

    # Create PDF with all plots
    print("Creating PDF report...")
    # Save PDF in the python folder (script directory)
    pdf_path = os.path.join(os.path.dirname(__file__), 'clustering_report.pdf')
    
    # Get all png files in visuals dir
    plot_files = [f for f in os.listdir(VISUALS_DIR) if f.endswith('.png')]
    
    # Sort them to have a logical order: distributions first, then others
    plot_files.sort(key=lambda x: (not x.startswith('dist_'), x))

    with PdfPages(pdf_path) as pdf:
        for filename in plot_files:
            filepath = os.path.join(VISUALS_DIR, filename)
            # Read image
            img = plt.imread(filepath)
            
            # Create a figure to display the image
            plt.figure(figsize=(12, 8))
            plt.imshow(img)
            plt.axis('off') # Hide axes
            plt.title(filename, fontsize=10)
            
            # Save to PDF
            pdf.savefig()
            plt.close()
            
    print(f"PDF report saved to: {pdf_path}")

if __name__ == "__main__":
    main()

File: R/clustering.R
# Load necessary libraries
library(tidyverse)
library(cluster)
library(factoextra)
library(mclust)

# Create visuals directory
if (!dir.exists("visuals")) {
  dir.create("visuals")
}

save_plot <- function(filename, plot_obj) {
  ggsave(file.path("visuals", filename), plot = plot_obj, width = 10, height = 6)

}

print("Loading dataset...")
df <- read.csv("../project_cluster.csv")

print("Preprocessing data...")

df <- df %>%
  mutate(
    Total_Guests = number.of.adults + number.of.children,
    Is_Family = as.integer(number.of.children > 0),
    Total_Nights = number.of.weekend.nights + number.of.week.nights,
    Cancellation_Ratio = (P.C + 1) / (P.not.C + P.C + 2),
    Price_per_Person = average.price / Total_Guests
  )

df$Price_per_Person[is.infinite(df$Price_per_Person)] <- NA

df <- df %>% select(-date.of.reservation)

df <- df %>%
  mutate(
    number.of.children = ifelse(number.of.children == 0, 0, 1),
    number.of.weekend.nights = pmin(number.of.weekend.nights, 3),
    number.of.week.nights = pmin(number.of.week.nights, 6),
    P.C = ifelse(P.C == 0, 0, 1),
    P.not.C = ifelse(P.not.C == 0, 0, 1),
    Total_Guests = ifelse(Total_Guests == 0, 1, pmin(Total_Guests, 3)),
    Total_Nights = pmin(round(Total_Nights), 8)
  )

# Save booking status and drop unused columns
booking_status <- df$booking.status
df <- df %>% select(-Booking_ID, -booking.status)

# One-hot encoding for market.segment.type
# R handles factors automatically in many cases, but for clustering we need numeric
# We'll use model.matrix to create dummy variables
dummy <- model.matrix(~ market.segment.type - 1, data = df)
df_encoded <- cbind(df %>% select(-market.segment.type), dummy)

# Drop meal, room type, 
df_encoded <- df_encoded %>% select(-starts_with("type.of.meal"), -starts_with("room.type"))

# Fill NaNs (with mean)
df_encoded <- df_encoded %>%
  mutate(across(everything(), ~ ifelse(is.na(.), mean(., na.rm = TRUE), .)))

# Scaling (MinMax -1 to 1)
# Function for MinMax scaling to [-1, 1]
minmax_scale <- function(x) {
  min_x <- min(x)
  max_x <- max(x)
  return(2 * ((x - min_x) / (max_x - min_x)) - 1)
}

df_scaled <- df_encoded %>%
  mutate(across(everything(), minmax_scale))

print("Data processed and scaled.")

# --- Plotting Distributions ---
print("Generating distribution plots...")
numeric_cols <- c(
  'Price_per_Person', 'Cancellation_Ratio', 'Total_Nights', 'Is_Family', 
  'Total_Guests', 'special.requests', 'average.price', 'P.not.C', 'P.C', 
  'repeated', 'lead.time', 'car.parking.space', 'number.of.week.nights', 
  'number.of.weekend.nights', 'number.of.children', 'number.of.adults'
)

for (col in numeric_cols) {
  if (col %in% names(df_scaled)) {
    p <- ggplot(df_scaled, aes(x = .data[[col]])) +
      geom_histogram(bins = 30, fill = "steelblue", color = "black", alpha = 0.7) +
      labs(title = paste("Distribution (Scaled):", col), x = "Scaled Value [-1, 1]", y = "Frequency") +
      theme_minimal() +
      theme(panel.grid.major.y = element_line(colour = "grey80"),
            panel.grid.major.x = element_blank(),
            panel.grid.minor = element_blank())

    # Removed print(p) to avoid Rplots.pdf
    save_plot(paste0("dist_", col, ".png"), p)
  }
}

# --- Clustering ---
print("Running clustering algorithms...")
library(gridExtra)  

X <- as.matrix(df_scaled)
k_range <- 2:10

kmeans_silhouette <- numeric(length(k_range))
hierarchical_silhouette <- numeric(length(k_range))
nmi_scores <- numeric(length(k_range))
ari_scores <- numeric(length(k_range))
kmeans_inertia <- numeric(length(k_range))

kmeans_labels_list <- list()
hierarchical_labels_list <- list()

for (i in seq_along(k_range)) {
  k <- k_range[i]
  
  # K-Means
  km <- kmeans(X, centers = k, nstart = 10)
  kmeans_labels_list[[i]] <- km$cluster
  kmeans_inertia[i] <- km$tot.withinss
  
  # Silhouette 
  ss_km <- silhouette(km$cluster, dist(X))
  kmeans_silhouette[i] <- mean(ss_km[, 3])
  
  # Hierarchical
  
  set.seed(42)
  idx <- sample(nrow(X), min(5000, nrow(X)))
  X_sub <- X[idx, ]
  
  hc <- hclust(dist(X_sub), method = "ward.D2")
  h_labels <- cutree(hc, k = k)
  
  ss_hc <- silhouette(h_labels, dist(X_sub))
  hierarchical_silhouette[i] <- mean(ss_hc[, 3])
  
  km_labels_sub <- kmeans_labels_list[[i]][idx]
  
  # Calculate NMI and ARI
  if (requireNamespace("aricode", quietly = TRUE)) {
    nmi_scores[i] <- aricode::NMI(km_labels_sub, h_labels)
  } else {
    nmi_scores[i] <- adjustedRandIndex(km_labels_sub, h_labels)
  }
  ari_scores[i] <- adjustedRandIndex(km_labels_sub, h_labels)
}

# Evaluation Plot - Combined
eval_df <- data.frame(
  k = k_range,
  Inertia = kmeans_inertia,
  Silhouette_KMeans = kmeans_silhouette,
  Silhouette_Hierarchical = hierarchical_silhouette,
  NMI = nmi_scores,
  ARI = ari_scores
)

# Create 4 subplots in 2x2 grid
p1 <- ggplot(eval_df, aes(x = k, y = Inertia)) + 
  geom_line() + 
  geom_point() + 
  labs(title = "K-Means Inertia", x = "Number of Clusters (k)", y = "Inertia") +
  theme_minimal()

p2 <- ggplot(eval_df, aes(x = k)) + 
  geom_line(aes(y = Silhouette_KMeans, color = "K-Means")) + 
  geom_point(aes(y = Silhouette_KMeans, color = "K-Means")) +
  geom_line(aes(y = Silhouette_Hierarchical, color = "Hierarchical")) + 
  geom_point(aes(y = Silhouette_Hierarchical, color = "Hierarchical")) +
  labs(title = "Silhouette Score", x = "Number of Clusters (k)", y = "Silhouette Score", color = "Method") +
  theme_minimal()

p3 <- ggplot(eval_df, aes(x = k, y = NMI)) + 
  geom_line() + 
  geom_point() + 
  labs(title = "NMI Score", x = "Number of Clusters (k)", y = "NMI") +
  theme_minimal()

p4 <- ggplot(eval_df, aes(x = k, y = ARI)) + 
  geom_line() + 
  geom_point() + 
  labs(title = "ARI Score", x = "Number of Clusters (k)", y = "ARI") +
  theme_minimal()

combined_plot <- grid.arrange(p1, p2, p3, p4, ncol = 2, nrow = 2)
ggsave("visuals/clustering_evaluation.png", combined_plot, width = 15, height = 12)

# --- Final Clustering (k=4) ---
print("Finalizing clusters (k=4)...")
k_optimal <- 4
set.seed(42)
km_final <- kmeans(X, centers = k_optimal, nstart = 10)
df_encoded$Cluster <- as.factor(km_final$cluster - 1)
df_encoded$booking.status <- booking_status

cluster_means <- df_encoded %>%
  group_by(Cluster) %>%
  summarise(across(where(is.numeric), mean))

print("Cluster Summary (Means):")
print(cluster_means)

print("Creating cluster comparison plots...")
variables_to_analyze <- c(
  'number.of.adults', 'number.of.children', 'car.parking.space', 
  'lead.time', 'repeated', 'P.C', 'P.not.C', 'average.price', 
  'special.requests', 'Total_Guests', 'Is_Family', 'Total_Nights', 
  'Cancellation_Ratio', 'Price_per_Person'
)

market_segment_cols <- names(df_encoded)[grepl("market.segment.type", names(df_encoded))]

all_vars <- c(variables_to_analyze, market_segment_cols)

cluster_colors <- c("steelblue", "coral", "lightgreen", "gold")

for (var in all_vars) {
  if (var %in% names(cluster_means)) {
    plot_data <- cluster_means %>%
      select(Cluster, all_of(var)) %>%
      rename(value = all_of(var))
    
    p <- ggplot(plot_data, aes(x = Cluster, y = value, fill = Cluster)) +
      geom_bar(stat = "identity", color = "black", alpha = 0.8) +
      scale_fill_manual(values = cluster_colors) +
      labs(title = paste("Cluster Comparison:", var), 
           x = "Cluster", 
           y = paste("Mean (", var, ")", sep = "")) +
      theme_minimal() +
      theme(legend.position = "none",
            panel.grid.major.y = element_line(colour = "grey80"),
            panel.grid.major.x = element_blank(),
            panel.grid.minor = element_blank())
    
    ggsave(file.path("visuals", paste0("cluster_comp_", gsub("\\\\.", "_", var), ".png")), 
           p, width = 10, height = 6)
  }
}

print("Cluster comparison plots created.")

cluster_summary <- cluster_means %>%
  mutate(Count = table(df_encoded$Cluster)[as.character(Cluster)]) %>%
  select(Count, everything(), -Cluster)
print("Creating heatmap...")
# Normalize cluster_summary for heatmap
cluster_summary_matrix <- as.matrix(cluster_summary)
cluster_summary_norm <- apply(cluster_summary_matrix, 2, function(x) {
  (x - min(x)) / (max(x) - min(x))
})
rownames(cluster_summary_norm) <- levels(df_encoded$Cluster)

heatmap_data <- as.data.frame(cluster_summary_norm) %>%
  mutate(Cluster = rownames(.)) %>%
  pivot_longer(-Cluster, names_to = "Variable", values_to = "Value")

p_heatmap <- ggplot(heatmap_data, aes(x = Cluster, y = Variable, fill = Value)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%.2f", Value)), color = "black", size = 3) +  # Add annotations
  scale_fill_gradient2(low = "red", mid = "yellow", high = "green", midpoint = 0.5, name = "Normalized\nValue (0-1)") +
  labs(title = "Cluster Means Heatmap (Normalized)", x = "Cluster", y = "Variable") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))

ggsave("visuals/cluster_heatmap.png", p_heatmap, width = 16, height = 6)
print("Heatmap created.")

cancel_rates <- df_encoded %>%
  group_by(Cluster) %>%
  summarise(Cancellation_Rate = mean(booking.status == "Canceled") * 100)

print("Cancellation Rates:")
print(cancel_rates)

p_cancel <- ggplot(cancel_rates, aes(x = Cluster, y = Cancellation_Rate, fill = Cluster)) +
  geom_bar(stat = "identity", color = "black", alpha = 0.8) +
  scale_fill_manual(values = c("steelblue", "coral", "lightgreen", "gold")) +
  labs(title = "Cancellation Rate by Cluster", x = "Cluster", y = "Cancellation Rate (%)") +
  theme_minimal() +
  theme(legend.position = "none",
        panel.grid.major.y = element_line(colour = "grey80"),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank())

save_plot("cancellation_rate.png", p_cancel)

# Create PDF report with all plots
print("Creating PDF report...")
library(magick)
library(grid)
library(gridExtra)

pdf_path <- "clustering_report.pdf"

plot_files <- list.files("visuals", pattern = "\\.png$", full.names = FALSE)
# Sort them to have a logical order: distributions first, then others
plot_files <- plot_files[order(!grepl("^dist_", plot_files), plot_files)]

pdf(pdf_path, width = 12, height = 8)

for (filename in plot_files) {
  filepath <- file.path("visuals", filename)
  
  img <- image_read(filepath)
  
  grid.newpage()
  grid.raster(img)
  
  grid.text(filename, x = 0.5, y = 0.98, 
            gp = gpar(fontsize = 10, col = "black"))
}

dev.off()

cat(paste("PDF report saved to:", pdf_path, "\n"))
print("R script completed successfully.")
